apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
+                            _extendQueue.Enqueue(msgIds.Select(x => new TimedId(_extendThrottleHigh + 1, x)));
                         }
-                        // Queue lease extensions on all not-yet-processed messages.
-                        // It's OK if an ID is lease-extended and ACKed to the server concurrently.
-                        // The ACK takes precedence on the server.
-                        QueueLeaseExtensions(ackIds);
                     }
                 }
-                finally
+                if (anyMsgIds)
                 {
-                    allMsgsHandled.Dispose();
+                    // Ids have been added to _extendQueue, so trigger a push.
+                    _eventPush.Set();
+                    // Some ids still exist, schedule another extension.
+                    Add(_scheduler.Delay(_autoExtendInterval, _softStopCts.Token), Next(false, () => HandleExtendLease(msgIds)));
+                    // Increment _extendThrottles.
+                    _extendThrottleHigh += 1;
+                    Add(_scheduler.Delay(_extendQueueThrottleInterval, _softStopCts.Token), Next(false, () => _extendThrottleLow += 1));
                 }
             }
 
-            private async Task Push(SubscriberServiceApiClient.StreamingPullStream pull, CancellationToken streamingPullToken)
+            private void HandlePush()
             {
-                // Send ack/extends in smaller chunks than maximum ack/extend queue size.
-                // This temporally smooths out server message receipt, which causes higher
-                // sustained message delivery in high-bandwidth environments.
-                var maxAckExtendSendCount = Math.Max(10, _maxAckExtendQueueSize / 4);
-                // Pushing of acks and extends
-                // If a Push call fails then this Task always ends, and the StreamingPull is cancelled.
-                // If it's recoverable, then a new StreamingPull call is started.
-                // If it's unrecoverable, then a global hard stop occurs.
-                while (true)
+                // Always re-listen for push events.
+                Add(_eventPush.WaitAsync(_pushStopCts.Token), Next(false, HandlePush));
+                // Send data, if there is any to send.
+                StartPush();
+            }
+
+            private void StartPush()
+            {
+                // Send data, if there is any to send.
+                if (_concurrentPushCount >= _maxConcurrentPush)
+                {
+                    // Too many existing concurrent pushes; do nothing.
+                    return;
+                }
+                List<string> acks;
+                List<TimedId> extends;
+                List<string> nacks;
+                lock (_lock)
+                {
+                    // Priority of sending: Acks, Extends, Nacks.
+                    // Allow 2 over _maxConcurrentPush, otherwise extend/nack queues could be unfairly penalized.
```

Comment:
It's not clear what this comment line has to do with the code after it.

Reply:
Improved comment.